### DISCUSSION TITLE
How to bind?

### DISCUSSION THREAD

**Discussion by PlayCreatively** [CommunityToolkit/Maui.Markup#74] [Q&A] [answered]
I was trying to bind a label's `TextProperty` to the page's local string (just to see how binding works and get the ball rolling), but I'm not able to make it bind.

This is the code I've written and I've tried all kinds of different variations of this but non work, so clearly I don't understand how binding works. However, I haven't found any info on the internet on how to make this simple binding. I've only seen people use an event handler or the sample code [here](https://docs.microsoft.com/en-us/dotnet/communitytoolkit/maui/markup/extensions/bindable-object-extensions#explicit-property) which I'm trying to immitate, but evidently failing. Can someone explain this for me? 🙏

```
class MyPage : ContentPage
    {
        public string myVal = "initial value";
        public DataPage()
        {
            Content =
                new VerticalStackLayout()
                {
                    Children = 
                    { 
                        new Label().Bind(Label.TextProperty, path: nameof(MyPage.myVal), source: this)
                    }
                };
        }
    }
```

**Comment by bijington** [comment, marked as answer]
Hey @PlayCreatively,

We do link over to the .NET MAUI docs but I don't actually see anything clearly helpful there either. Essentially when creating a binding we need to do so against a **property** that is accessible (`public` or `internal`), to tell the control the context of the binding (`BindingContext`) and we also need a way of notifying the control when things change (`INotifyPropertyChanged`).

What you had was close but there are some extra bits you needed. See this modified version of your code:

```csharp
class MyPage : ContentPage
{
    public string MyVal { get; } = "initial value";

    public DataPage()
    {
        BindingContext = this;

        Content =
            new VerticalStackLayout()
            {
                Children = 
                { 
                    new Label().Bind(Label.TextProperty, path: nameof(MyPage.MyVal))
                }
            };
    }
}
```

The changes I made are:

1. Set the `BindingContext` on the page to itself to know where the binding comes from.
1. Changed your field `myVar` to a property `MyVar`
1. I haven't added the notification of changes as you don't need it in this example and it adds more noise to getting this sample working. I can add that separately if you need

**Reply by PlayCreatively** [reply]
> I haven't added the notification of changes as you don't need it in this example and it adds more noise to getting this sample working. I can add that separately if you need

I would actually love it if you could maybe bind the `Label` to an `Entry` so I can see how it works.

How I understand how it works is that when a `View` has an `INotifyPropertyChanged` interface the XAML subscribes to it automatically when you bind it through XAML, but since I'm not using XAML I'll have to do it manually, I'm guessing 🤔

**Reply by bijington** [reply]
Oh if you are binding two controls together you can actually ignore my point about `INotifyPropertyChanged` we use this when one side of the binding needs to trigger the updates which is typically implemented in a View Model, the controls themselves already handle the updates for you.

If you wanted to bind your `Label` to an `Entry` you can do something like:

```csharp
class MyPage : ContentPage
{
    public DataPage()
    {
        Content =
            new VerticalStackLayout()
            {
                Children = 
                { 
                    new Entry().Assign(out var myEntry),

                    new Label()
                      .Bind(
                        Label.TextProperty,
                        path: nameof(Entry.Text),
                        source: myEntry)
                }
            };
    }
}
```

This is binding the `TextProperty` of your `Label` to the `Text` property on your `Entry` called `myEntry`.

The `Assign` method is another extension that we provide to allow you to keep creating the UI fluently but assign it out to a variable for use later on, [more documentation here](https://docs.microsoft.com/dotnet/communitytoolkit/maui/markup/extensions/bindable-object-extensions#assign)

**Comment by CliffAgius** [comment]
Closed and answered
